-Ysafe-init crashes instead of reporting indirect initialization error

## Compiler version

If you're not sure what version you're using, run `print scalaVersion` from sbt
(if you're running scalac manually, use `scalac -version` instead).

3.2.1

## Minimized code
https://scastie.scala-lang.org/RfKNWVLgTtaa2M55IlyJ8g  <--- crashes (see output below
https://scastie.scala-lang.org/QtU8dx5WS1Cv9AJkBFlYlQ <--- works as expected (-Ysafe-init disabled in build settings)
```
parameterized trait ATrait is indirectly implemented,
needs to be implemented directly so that arguments can be passed
```

## Output (click arrow to expand)
<details>
```
java.lang.AssertionError: assertion failed: arguments = 0, params = 2
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.transform.init.Semantic$Env$.apply(Semantic.scala:175)
	at dotty.tools.dotc.transform.init.Semantic$.given_Env$lzyINIT2$1(Semantic.scala:758)
	at dotty.tools.dotc.transform.init.Semantic$.given_Env$2(Semantic.scala:758)
	at dotty.tools.dotc.transform.init.Semantic$.$anonfun$15(Semantic.scala:761)
	at dotty.tools.dotc.transform.init.Semantic$.callConstructor(Semantic.scala:761)
	at dotty.tools.dotc.transform.init.Semantic$.superCall$1$$anonfun$1(Semantic.scala:1434)
	at dotty.tools.dotc.transform.init.Semantic$.init$$anonfun$3(Semantic.scala:1496)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:926)
	at dotty.tools.dotc.transform.init.Semantic$.init(Semantic.scala:1496)
	at dotty.tools.dotc.transform.init.Semantic$.cases(Semantic.scala:1299)
	at dotty.tools.dotc.transform.init.Semantic$.eval$$anonfun$1(Semantic.scala:1110)
	at dotty.tools.dotc.transform.init.Semantic$Cache$Cache.assume(Semantic.scala:311)
	...
```
</details>

--- a/compiler/src/dotty/tools/dotc/transform/init/Semantic.scala
+++ b/compiler/src/dotty/tools/dotc/transform/init/Semantic.scala
@@ -855,7 +855,7 @@ object Semantic:
       // init "fake" param fields for parameters of primary and secondary constructors
       def addParamsAsFields(args: List[Value], ref: Ref, ctorDef: DefDef) =
         val params = ctorDef.termParamss.flatten.map(_.symbol)
-        assert(args.size == params.size, "arguments = " + args.size + ", params = " + params.size + ", ctro = " + ctor.show)
+        assert(args.size == params.size, "arguments = " + args.size + ", params = " + params.size + ", ctor = " + ctor.show)
         for (param, value) <- params.zip(args) do
           ref.updateField(param, value)
           printer.println(param.show + " initialized with " + value)

--- a/compiler/src/dotty/tools/dotc/transform/init/Semantic.scala
+++ b/compiler/src/dotty/tools/dotc/transform/init/Semantic.scala
@@ -855,7 +855,7 @@ object Semantic:
       // init "fake" param fields for parameters of primary and secondary constructors
       def addParamsAsFields(args: List[Value], ref: Ref, ctorDef: DefDef) =
         val params = ctorDef.termParamss.flatten.map(_.symbol)
-        assert(args.size == params.size, "arguments = " + args.size + ", params = " + params.size)
+        assert(args.size == params.size, "arguments = " + args.size + ", params = " + params.size + ", ctro = " + ctor.show)
         for (param, value) <- params.zip(args) do
           ref.updateField(param, value)
           printer.println(param.show + " initialized with " + value)
@@ -1663,12 +1663,13 @@ object Semantic:
           // term arguments to B. That can only be done in a concrete class.
           val tref = typeRefOf(klass.typeRef.baseType(mixin).typeConstructor)
           val ctor = tref.classSymbol.primaryConstructor
-          if ctor.exists && ctor.paramSymss.isEmpty then
+          if ctor.exists then
             // The parameter check of traits comes late in the mixin phase.
-            // To avoid crash we ignore the initialization check for erroneous
-            // parent call code. See tests/neg/i16438.scala.
+            // To avoid crash we supply hot values for erroneous parent calls.
+            // See tests/neg/i16438.scala.
+            val args: List[ArgInfo] = ctor.info.paramInfoss.flatten.map(_ => ArgInfo(Hot, Trace.empty))
             extendTrace(superParent) {
-              superCall(tref, ctor, args = Nil, tasks)
+              superCall(tref, ctor, args, tasks)
             }
       }
 

--- a/compiler/src/dotty/tools/dotc/transform/init/Semantic.scala
+++ b/compiler/src/dotty/tools/dotc/transform/init/Semantic.scala
@@ -1663,9 +1663,13 @@ object Semantic:
           // term arguments to B. That can only be done in a concrete class.
           val tref = typeRefOf(klass.typeRef.baseType(mixin).typeConstructor)
           val ctor = tref.classSymbol.primaryConstructor
-          if ctor.exists then extendTrace(superParent) {
-            superCall(tref, ctor, Nil, tasks)
-          }
+          if ctor.exists && ctor.paramSymss.isEmpty then
+            // The parameter check of traits comes late in the mixin phase.
+            // To avoid crash we ignore the initialization check for erroneous
+            // parent call code. See tests/neg/i16438.scala.
+            extendTrace(superParent) {
+              superCall(tref, ctor, args = Nil, tasks)
+            }
       }
 
       // initialize super classes after outers are set

--- a/tests/neg/i16438.scala
+++ b/tests/neg/i16438.scala
@@ -1,0 +1,4 @@
+// scalac: -Ysafe-init
+trait ATrait(val string: String, val int: Int)
+trait AnotherTrait( override val string: String, override val int: Int) extends ATrait
+case class ACaseClass(override val string: String) extends AnotherTrait(string, 3) // error

